lib/ansible/module_utils/network/ftd/configuration.py

```diff
@@ -28,86 +30,358 @@
+        self._check_mode = check_mode
+        self._operation_checker = OperationChecker
 
-    def get_object_by_name(self, url_path, name, path_params=None):
-        item_generator = iterate_over_pageable_resource(
-            partial(self.send_request, url_path=url_path, http_method=HTTPMethod.GET, path_params=path_params),
-            {'filter': 'name:%s' % name}
-        )
-        # not all endpoints support filtering so checking name explicitly
-        return next((item for item in item_generator if item['name'] == name), None)
+    def execute_operation(self, op_name, params):
+        """
+        Allow user request execution of simple operations(natively supported by API provider) as well as complex
+        operations(operations that are implemented as a set of simple operations).
+
+        :param op_name: name of the operation being called by the user
+        :type op_name: str
+        :param params: definition of the params that operation should be executed with
+        :type params: dict
+        :return: Result of the operation being executed
+        :rtype: dict
+        """
+        if self._operation_checker.is_upsert_operation(op_name):
+            return self.upsert_object(op_name, params)
+        else:
+            return self.crud_operation(op_name, params)
 
-    def get_objects_by_filter(self, url_path, filters, path_params=None, query_params=None):
-        def match_filters(obj):
-            for k, v in filters.items():
+    def crud_operation(self, op_name, params):
+        """
+        Allow user request execution of simple operations(natively supported by API provider) only.
+
+        :param op_name: name of the operation being called by the user
+        :type op_name: str
+        :param params: definition of the params that operation should be executed with
+        :type params: dict
+        :return: Result of the operation being executed
+        :rtype: dict
+        """
+        op_spec = self.get_operation_spec(op_name)
+        if op_spec is None:
+            raise FtdInvalidOperationNameError(op_name)
+
+        if self._operation_checker.is_add_operation(op_name, op_spec):
+            resp = self.add_object(op_name, params)
+        elif self._operation_checker.is_edit_operation(op_name, op_spec):
+            resp = self.edit_object(op_name, params)
+        elif self._operation_checker.is_delete_operation(op_name, op_spec):
+            resp = self.delete_object(op_name, params)
+        elif self._operation_checker.is_find_by_filter_operation(op_name, params, op_spec):
+            resp = list(self.get_objects_by_filter(op_name, params))
+        else:
+            resp = self.send_general_request(op_name, params)
+        return resp
+
+    def get_operation_spec(self, operation_name):
+        if operation_name not in self._operation_spec_cache:
+            self._operation_spec_cache[operation_name] = self._conn.get_operation_spec(operation_name)
```

Comment:
on py2.7 I am getting below error when trying to add new Network_object
```
The full traceback is:
Traceback (most recent call last):
  File "/home/vagrant/.ansible/tmp/ansible-local-26093iZ9x1I/ansible-tmp-1540995434.55-165674802918250/AnsiballZ_ftd_configuration.py", line 113, in <module>
    _ansiballz_main()
  File "/home/vagrant/.ansible/tmp/ansible-local-26093iZ9x1I/ansible-tmp-1540995434.55-165674802918250/AnsiballZ_ftd_configuration.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/vagrant/.ansible/tmp/ansible-local-26093iZ9x1I/ansible-tmp-1540995434.55-165674802918250/AnsiballZ_ftd_configuration.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_ftd_configuration_payload_3gzquE/__main__.py", line 140, in <module>
  File "/tmp/ansible_ftd_configuration_payload_3gzquE/__main__.py", line 121, in main
  File "/tmp/ansible_ftd_configuration_payload_3gzquE/ansible_ftd_configuration_payload.zip/ansible/module_utils/network/ftd/configuration.py", line 222, in execute_operation
  File "/tmp/ansible_ftd_configuration_payload_3gzquE/ansible_ftd_configuration_payload.zip/ansible/module_utils/network/ftd/configuration.py", line 235, in crud_operation
  File "/tmp/ansible_ftd_configuration_payload_3gzquE/ansible_ftd_configuration_payload.zip/ansible/module_utils/network/ftd/configuration.py", line 253, in get_operation_spec
  File "/tmp/ansible_ftd_configuration_payload_3gzquE/ansible_ftd_configuration_payload.zip/ansible/module_utils/connection.py", line 201, in __rpc__
ansible.module_utils.connection.ConnectionError: 'ascii' codec can't encode character u'\u2264' in position 190520: ordinal not in range(128)

```
my playbook has operation 
operation: "addNetworkObject"

Reply:
Good catch! It happened because `format` method does not handle unicode strings gracefully in Python 2.7 (more info [here](https://www.saltycrane.com/blog/2014/07/old-string-formatting-and-new-format-method-handle-unicode-differently/)). I updated this line to use the old string formatting that works both with unicode and byte strings.
We completely missed this issue as the API spec on real back-end does not contain unicode-specific characters, so all tests passed. But good that you found it 👍

Reply:
@annikulin new error. not sure how it is working for you . It it trying to index into list with a string  at last function in traceback
The full traceback is:
Traceback (most recent call last):
  File "/home/vagrant/.ansible/tmp/ansible-local-31309R9mPLw/ansible-tmp-1541166190.87-33859762983323/AnsiballZ_ftd_configuration.py", line 113, in <module>
    _ansiballz_main()
  File "/home/vagrant/.ansible/tmp/ansible-local-31309R9mPLw/ansible-tmp-1541166190.87-33859762983323/AnsiballZ_ftd_configuration.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/vagrant/.ansible/tmp/ansible-local-31309R9mPLw/ansible-tmp-1541166190.87-33859762983323/AnsiballZ_ftd_configuration.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_ftd_configuration_payload_EvFz6v/__main__.py", line 140, in <module>
  File "/tmp/ansible_ftd_configuration_payload_EvFz6v/__main__.py", line 121, in main
  File "/tmp/ansible_ftd_configuration_payload_EvFz6v/ansible_ftd_configuration_payload.zip/ansible/module_utils/network/ftd/configuration.py", line 223, in execute_operation
  File "/tmp/ansible_ftd_configuration_payload_EvFz6v/ansible_ftd_configuration_payload.zip/ansible/module_utils/network/ftd/configuration.py", line 240, in crud_operation
  File "/tmp/ansible_ftd_configuration_payload_EvFz6v/ansible_ftd_configuration_payload.zip/ansible/module_utils/network/ftd/configuration.py", line 84, in is_add_operation
  File "/tmp/ansible_ftd_configuration_payload_EvFz6v/ansible_ftd_configuration_payload.zip/ansible/module_utils/network/ftd/configuration.py", line 486, in is_post_request
TypeError: string indices must be integers

Reply:
yes this passes with latest FW v 6.3.0. In ansible module code, can we identify if api spec returned by device does not work  ? Do we have instances of cisco FTD v < 6.3.0 in field ?

Reply:
Yes, there is v6.2.3. We have tested the code with all FTD versions (including v6.2.3 and v6.3) and the [test playbooks](https://github.com/CiscoDevNet/FTDAnsible/tree/master/samples) passed.
We are also working on setting up a CI for [FTD Ansible](https://github.com/CiscoDevNet/FTDAnsible) repo to run playbooks with all FTD versions automatically.
The reason why it failed was the invalid API specs inside the demo app. For regular builds, it's being built during the build process while for the demo app it was included inside and this issue occurred.